README.md
# Backtrade

<p align="center">
  <a href="https://github.com/34j/backtrade/actions?query=workflow%3ACI">
    <img src="https://img.shields.io/github/workflow/status/34j/backtrade/CI/main?label=CI&logo=github&style=flat-square" alt="CI Status" >
  </a>
  <a href="https://backtrade.readthedocs.io">
    <img src="https://img.shields.io/readthedocs/backtrade.svg?logo=read-the-docs&logoColor=fff&style=flat-square" alt="Documentation Status">
  </a>
  <a href="https://codecov.io/gh/34j/backtrade">
    <img src="https://img.shields.io/codecov/c/github/34j/backtrade.svg?logo=codecov&logoColor=fff&style=flat-square" alt="Test coverage percentage">
  </a>
</p>
<p align="center">
  <a href="https://python-poetry.org/">
    <img src="https://img.shields.io/badge/packaging-poetry-299bd7?style=flat-square&logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAA4AAAASCAYAAABrXO8xAAAACXBIWXMAAAsTAAALEwEAmpwYAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAJJSURBVHgBfZLPa1NBEMe/s7tNXoxW1KJQKaUHkXhQvHgW6UHQQ09CBS/6V3hKc/AP8CqCrUcpmop3Cx48eDB4yEECjVQrlZb80CRN8t6OM/teagVxYZi38+Yz853dJbzoMV3MM8cJUcLMSUKIE8AzQ2PieZzFxEJOHMOgMQQ+dUgSAckNXhapU/NMhDSWLs1B24A8sO1xrN4NECkcAC9ASkiIJc6k5TRiUDPhnyMMdhKc+Zx19l6SgyeW76BEONY9exVQMzKExGKwwPsCzza7KGSSWRWEQhyEaDXp6ZHEr416ygbiKYOd7TEWvvcQIeusHYMJGhTwF9y7sGnSwaWyFAiyoxzqW0PM/RjghPxF2pWReAowTEXnDh0xgcLs8l2YQmOrj3N7ByiqEoH0cARs4u78WgAVkoEDIDoOi3AkcLOHU60RIg5wC4ZuTC7FaHKQm8Hq1fQuSOBvX/sodmNJSB5geaF5CPIkUeecdMxieoRO5jz9bheL6/tXjrwCyX/UYBUcjCaWHljx1xiX6z9xEjkYAzbGVnB8pvLmyXm9ep+W8CmsSHQQY77Zx1zboxAV0w7ybMhQmfqdmmw3nEp1I0Z+FGO6M8LZdoyZnuzzBdjISicKRnpxzI9fPb+0oYXsNdyi+d3h9bm9MWYHFtPeIZfLwzmFDKy1ai3p+PDls1Llz4yyFpferxjnyjJDSEy9CaCx5m2cJPerq6Xm34eTrZt3PqxYO1XOwDYZrFlH1fWnpU38Y9HRze3lj0vOujZcXKuuXm3jP+s3KbZVra7y2EAAAAAASUVORK5CYII=" alt="Poetry">
  </a>
  <a href="https://github.com/ambv/black">
    <img src="https://img.shields.io/badge/code%20style-black-000000.svg?style=flat-square" alt="black">
  </a>
  <a href="https://github.com/pre-commit/pre-commit">
    <img src="https://img.shields.io/badge/pre--commit-enabled-brightgreen?logo=pre-commit&logoColor=white&style=flat-square" alt="pre-commit">
  </a>
</p>
<p align="center">
  <a href="https://pypi.org/project/backtrade/">
    <img src="https://img.shields.io/pypi/v/backtrade.svg?logo=python&logoColor=fff&style=flat-square" alt="PyPI Version">
  </a>
  <img src="https://img.shields.io/pypi/pyversions/backtrade.svg?style=flat-square&logo=python&amp;logoColor=fff" alt="Supported Python versions">
  <img src="https://img.shields.io/pypi/l/backtrade.svg?style=flat-square" alt="License">
</p>

A small Python package for backtesting.

## Installation

Install this via pip (or your favourite package manager):

`pip install backtrade`

## Usage

```python
from backtrade import Backtester, CloseData, LimitOrder, MarketOrder, _IndexType

# Define my strategy
class MyBacktest(Backtester):
    def init(self):
        pass

    def next(
        self, close_data: "CloseData[_IndexType]", row: "Series[Any]"
    ) -> "Iterable[Union[MarketOrder, LimitOrder]]":
        yield MarketOrder(size=1 / close_data.close)
        yield MarketOrder(size=-1 / close_data.close)

# Run backtest and plot results
bt = MyBacktest()
bt(df, maker_fee=-0.025 * 0.01, taker_fee=0.001).plot()
```

## Contributors ✨

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- prettier-ignore-start -->
<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tbody>
    <tr>
      <td align="center"><a href="https://github.com/34j"><img src="https://avatars.githubusercontent.com/u/55338215?v=4?s=80" width="80px;" alt="34j"/><br /><sub><b>34j</b></sub></a><br /><a href="https://github.com/34j/backtrade/commits?author=34j" title="Code">💻</a></td>
    </tr>
  </tbody>
</table>

<!-- markdownlint-restore -->
<!-- prettier-ignore-end -->

<!-- ALL-CONTRIBUTORS-LIST:END -->
<!-- prettier-ignore-end -->

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification. Contributions of any kind welcome!

## Credits

This package was created with
[Copier](https://copier.readthedocs.io/) and the
[browniebroke/pypackage-template](https://github.com/browniebroke/pypackage-template)
project template.


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.7.2 (2022-12-02)
### Fix
* Fix typo ([#23](https://github.com/34j/backtrade/issues/23)) ([`3293d5b`](https://github.com/34j/backtrade/commit/3293d5be0ef2b91888d9e707ff6378a9c37a294c))

## v0.7.1 (2022-12-02)
### Fix
* Fix OrderBase typing ([#22](https://github.com/34j/backtrade/issues/22)) ([`7db5e20`](https://github.com/34j/backtrade/commit/7db5e204a43073e86427e1a5d7589d2676ff5087))

## v0.7.0 (2022-12-02)
### Feature
* Add __add__ and __mul__ for results ([#21](https://github.com/34j/backtrade/issues/21)) ([`04a2c71`](https://github.com/34j/backtrade/commit/04a2c7164a93db815e0add43cf91428e5a959261))

## v0.6.0 (2022-12-01)
### Feature
* Show tqdm bar ([#20](https://github.com/34j/backtrade/issues/20)) ([`db30e8f`](https://github.com/34j/backtrade/commit/db30e8f143519bff778416437f9b7c36699855b9))

## v0.5.0 (2022-12-01)
### Feature
* Add more metrics ([#19](https://github.com/34j/backtrade/issues/19)) ([`a696a39`](https://github.com/34j/backtrade/commit/a696a3962b8463406d7a81630b1f903d3fa2c82f))

## v0.4.1 (2022-11-30)
### Fix
* Fix wrong fee calculation ([#18](https://github.com/34j/backtrade/issues/18)) ([`abfaa6a`](https://github.com/34j/backtrade/commit/abfaa6a4443f3373cd69aed18ebf3eac9ef701af))

## v0.4.0 (2022-11-30)
### Feature
* Add more metrics ([#16](https://github.com/34j/backtrade/issues/16)) ([`e6fb20b`](https://github.com/34j/backtrade/commit/e6fb20b9d7539fb2e78e382ee8aed758ee30e704))

## v0.3.1 (2022-11-29)
### Fix
* Fix drawdown calculation ([#14](https://github.com/34j/backtrade/issues/14)) ([`181aa7e`](https://github.com/34j/backtrade/commit/181aa7e9b90eb0bc1af29ca3fdddd9ea91367d2c))

## v0.3.0 (2022-11-29)
### Feature
* Add metrics ([#13](https://github.com/34j/backtrade/issues/13)) ([`05e6af6`](https://github.com/34j/backtrade/commit/05e6af6ded42f0f60fd22a38401200e715fbdc3d))

## v0.2.0 (2022-11-28)
### Feature
* Add validator to attrs class ([#12](https://github.com/34j/backtrade/issues/12)) ([`01622c1`](https://github.com/34j/backtrade/commit/01622c1e196017b208da015e19f5ad95d5984e9d))

## v0.1.0 (2022-11-26)
### Feature
* Trigger release ([`1001dc7`](https://github.com/34j/backtrade/commit/1001dc7b1eb5eb4d5d4f1a646b15835b596f9270))
* Add main feat ([`12caff1`](https://github.com/34j/backtrade/commit/12caff13ec5a3a9bd7a1d67d6adf7e46e8248afb))

### Fix
* Fix version ([`d2a1e8a`](https://github.com/34j/backtrade/commit/d2a1e8a1a3aae76310c4a221666345328a374247))
* Trigger release ([`9155ac1`](https://github.com/34j/backtrade/commit/9155ac156b7cbaaf893a0de2c7f5eb18c7198222))
* Trigger release ([`cdb1521`](https://github.com/34j/backtrade/commit/cdb152109d1a1f078e77deac897940b6d14755fc))
* Forgot rename ([`95e958d`](https://github.com/34j/backtrade/commit/95e958debf588019427f73bd8d11e8cf9d59d087))
* Support python3.8 type hints ([`2e35d78`](https://github.com/34j/backtrade/commit/2e35d789cd03b431e3f4d72774ed171300d11974))
* Fix wrong dtype for Series ([`2a41965`](https://github.com/34j/backtrade/commit/2a41965172910e0bebbda39081b24c0731910695))
* Fix pyproject.toml ([`a2b7551`](https://github.com/34j/backtrade/commit/a2b755141681ad2070164b9308aee831ca274af8))
* Import annotations from __future__ ([`bb62f6f`](https://github.com/34j/backtrade/commit/bb62f6f171a43031bfee13f5dcd71843b1717c8c))

### Documentation
* Add 34j as a contributor for code ([#6](https://github.com/34j/backtrade/issues/6)) ([`0d47a26`](https://github.com/34j/backtrade/commit/0d47a26e6d00e6d448978ad7e1fae0faa1768d4c))
* Add docs ([`21cdff3`](https://github.com/34j/backtrade/commit/21cdff36049d131d09aa48d83d2698ec55454844))
* Example ([`321bd2e`](https://github.com/34j/backtrade/commit/321bd2e1bb1d73c9abd4393cfe159e2fdf165960))
* Readme ([`7884901`](https://github.com/34j/backtrade/commit/7884901ed26cd4ffd4319620ce0ce9a71a85ab50))


src/backtrade/backtest/backtest.py
from __future__ import annotations

import warnings
from abc import ABCMeta, abstractmethod
from functools import reduce
from logging import getLogger
from typing import Any, Generic, Iterable, final

import attrs
import joblib
import numpy as np
from exceptiongroup import ExceptionGroup
from pandas import DataFrame, Series
from tqdm import tqdm

from backtrade.logic import FinishedOrder, ProcessOrderArgs, process_order

from ..order import LimitOrder, MarketOrder, _IndexType
from .dtypes import BacktestResult, CloseData

__all__ = ["Backtester"]


class Backtester(Generic[_IndexType], metaclass=ABCMeta):
    """Simple backtester.

    Supports hedge mode, different `maker` / `taker` fees,
    `market`, `limit`, `post-only limit` orders.

    Every unfilled orders are to be automatically canceled
    on the next time step.

    Example
    ----------
    >>> import pandas as pd
    >>> from backtrade import (Backtester, CloseData, LimitOrder,
            MarketOrder, _IndexType)
    >>> df = pd.concat([open, close, high, low, signal, ATR], axis=1)
    >>> class MyBacktester(Backtester[_IndexType]):
    >>>    def init(self):
    >>>        pass
    >>>    def on_close(
                self,
                close_data: "CloseData[_IndexType]",
                row: "Series[Any]"
            ) -> "Iterable[Union[MarketOrder, LimitOrder]]":
    >>>        if row["signal"] == 1:
    >>>            yield Order(size=1, price=row['close'] - row['ATR'] * 0.5,
                               limit=True, post_only=True)
    >>>        if row["signal"] == -1:
    >>>            yield Order(size=-1, price=row['close'] + row['ATR'] * 0.5,
                               limit=True, post_only=True)
    >>> bt = Backtester()
    >>> bt(df, maker_fee=-0.01 * 0.025, taker_fee=0.01 * 0.075).plot()

    """

    def _parrarel(
        self,
        df: DataFrame,
        *,
        maker_fee: float,
        taker_fee: float,
        balance_init: float = 1,
        name: str | None = None,
        n_splits: int = -1,
        logarithmic: bool = True,
    ) -> BacktestResult[_IndexType]:
        if n_splits == 0:
            raise ValueError("n_splits must be not 0")
        if n_splits < -joblib.parallel.cpu_count():
            raise ValueError(
                "n_splits must be greater than "
                + f"-cpu_count={-joblib.parallel.cpu_count()}"
            )
        if n_splits < 0:
            n_splits = joblib.parallel.cpu_count() + n_splits + 1

        df_split = np.array_split(df, n_splits)
        maker_fee_split = np.full(n_splits, maker_fee)
        taker_fee_split = np.full(n_splits, taker_fee)
        balance_init_split = np.full(n_splits, balance_init)
        name_split = np.full(n_splits, name)
        logarithmic_split = np.full(n_splits, logarithmic)
        results: list[BacktestResult[_IndexType]] | None = joblib.Parallel(
            n_jobs=-1, verbose=10
        )(
            joblib.delayed(self)(
                df,
                maker_fee=maker_fee,
                taker_fee=taker_fee,
                balance_init=balance_init,
                n_splits=1,
                logarithmic=logarithmic,
                name=name,
                use_tqdm=False,
            )
            for df, maker_fee, taker_fee, balance_init, logarithmic, name in zip(
                df_split,
                maker_fee_split,
                taker_fee_split,
                balance_init_split,
                logarithmic_split,
                name_split,
            )
        )
        if results is None:
            raise ValueError("Joblib returned None.")

        if logarithmic:
            result = reduce(
                lambda x, y: x + y * (x.equity_quote[-1] / y.equity_quote[0]), results
            )
            result = attrs.evolve(result, name=name)
        else:
            result = reduce(lambda x, y: x + y, results)
            result = attrs.evolve(
                result,
                name=name,
                equity_quote=result.equity_quote - balance_init * (n_splits - 1),
                balance_quote=result.balance_quote - balance_init * (n_splits - 1),
            )
        return result

    @final
    def __call__(
        self,
        df: DataFrame,
        *,
        maker_fee: float,
        taker_fee: float,
        balance_init: float = 1,
        name: str | None = None,
        n_splits: int = 1,
        logarithmic: bool = True,
        use_tqdm: bool = True,
    ) -> BacktestResult[_IndexType]:
        """Initialize the backtester.

        Parameters
        ----------
        df : DataFrame
            Must have columns 'open', 'close', 'high', 'low'.
            df.loc[index, :] will be passed to `next` method.
        maker_fee : float
            Maker fee.
        taker_fee : float
            Taker fee.
        balance_init : float, optional
            Initial balance, by default 1
        name : str, optional
            Name of the backtest, by default None
        """

        # Check arguments
        # Errors
        errors: list[ValueError] = []
        if not all(col in df.columns for col in ["open", "close", "high", "low"]):
            if all(col in df.columns for col in ["Open", "Close", "High", "Low"]):
                df = df.copy()
                df[["open", "close", "high", "low"]] = df[
                    ["Open", "Close", "High", "Low"]
                ]
            else:
                errors.append(
                    ValueError(
                        'df must have columns "open", "close", "high", "low", '
                        + f"but got {df.columns}"
                    )
                )
        if not errors:
            if (df["open"] > df["high"]).any():
                errors.append(ValueError("open price must be less than high price"))
            if (df["open"] < df["low"]).any():
                errors.append(ValueError("open price must be greater than low price"))
            if (df["close"] > df["high"]).any():
                errors.append(ValueError("close price must be less than high price"))
            if (df["close"] < df["low"]).any():
                errors.append(ValueError("close price must be greater than low price"))
            if (df["high"] < df["low"]).any():
                errors.append(ValueError("high price must be greater than low price"))
            if (df["open"] <= 0).any():
                errors.append(ValueError("open price must be greater than 0"))
            if (df["close"] <= 0).any():
                errors.append(ValueError("close price must be greater than 0"))
            if (df["high"] <= 0).any():
                errors.append(ValueError("high price must be greater than 0"))
            if (df["low"] <= 0).any():
                errors.append(ValueError("low price must be greater than 0"))
        if not df.index.is_monotonic_increasing:
            errors.append(ValueError("index must be monotonic increasing"))
        if not df.index.is_unique:
            errors.append(ValueError("index must be unique"))
        if not balance_init > 0:
            errors.append(ValueError("balance_init must be greater than 0"))

        # Warnings
        if not taker_fee > 0:
            warnings.warn(f"taker_fee is not positive (got {taker_fee}), are you sure?")

        # Raise errors
        if errors:
            raise ExceptionGroup(
                "Invalid arguments" + str([e.args[0] for e in errors]), errors
            )

        # Multiprocessing
        if n_splits != 1:
            return self._parrarel(
                df,
                maker_fee=maker_fee,
                taker_fee=taker_fee,
                balance_init=balance_init,
                name=name,
                n_splits=n_splits,
                logarithmic=logarithmic,
            )

        df = df.copy()
        # logger
        self.logger = getLogger(__name__)

        # Initialize
        self.init()

        # Calculate
        open_orders: tuple[LimitOrder | MarketOrder, ...] = ()
        last_close: float | None = None
        position = 0.0
        balance = balance_init

        position_history: Series[float] = Series(np.nan, index=df.index)
        position_quote_history: Series[float] = Series(np.nan, index=df.index)
        balance_quote_history: Series[float] = Series(np.nan, index=df.index)
        equity_quote_history: Series[float] = Series(np.nan, index=df.index)
        finished_orders_history: Series[list[FinishedOrder[_IndexType, Any]]] = Series(
            np.nan, index=df.index, dtype=object
        )

        pbar = tqdm(df.iterrows(), total=len(df), disable=not use_tqdm)
        for index, row in pbar:
            index: _IndexType  # type: ignore
            row: Series[float]  # type: ignore
            open_ = row["open"]
            high = row["high"]
            low = row["low"]

            # Assert
            # assert open_ >= low
            # assert open_ <= high
            # assert low > 0
            # assert open_ > 0
            # assert high > 0

            # Iterate each open orders
            finished_orders: list[FinishedOrder[_IndexType, Any]] = []
            for order in open_orders:
                assert last_close is not None  # nosec
                if order.size == 0.0:
                    continue
                finished_order = process_order(
                    ProcessOrderArgs(
                        order=order,
                        last_close=last_close,
                        high=high,
                        low=low,
                        maker_fee=maker_fee,
                        taker_fee=taker_fee,
                        index=index,
                    )
                )
                balance -= finished_order.balance_decrement
                if finished_order.filled:
                    position += order.size
                finished_orders.append(finished_order)

            # Call at close
            equity = balance + position * open_
            position_quote = position * open_
            open_orders = tuple(
                self.on_close(
                    CloseData(
                        index=index,
                        open=open_,
                        high=high,
                        low=low,
                        close=row["close"],
                        position=position,
                        position_quote=position_quote,
                        balance_quote=balance,
                        equity_quote=equity,
                    ),
                    row,
                )
            )
            last_close = row["close"]

            position_history[index] = position
            position_quote_history[index] = position_quote
            balance_quote_history[index] = balance
            equity_quote_history[index] = equity
            finished_orders_history[index] = finished_orders

        finished_orders_exploded = finished_orders_history.explode().dropna()

        return BacktestResult(
            name=name,
            close=df["close"],
            position=position_history.rename("position"),
            position_quote=position_quote_history.rename("position_quote"),
            balance_quote=balance_quote_history.rename("balance_quote"),
            equity_quote=equity_quote_history.rename("equity_quote"),
            finished_orders=finished_orders_exploded.rename("finished_orders"),
            maker_fee_rate=maker_fee,
            taker_fee_rate=taker_fee,
            logarithmic=logarithmic,
        )

    def init(self) -> None:
        """If you want to initialize something, override this method.
        Usually, you don't need to do this."""

    @abstractmethod
    def on_close(
        self, data: CloseData[_IndexType], row: Series[Any]
    ) -> Iterable[LimitOrder | MarketOrder]:
        """Override this method to implement your strategy.
        Unfilled orders would be automatically canceled.

        Parameters
        ----------
        data: CloseData
            CloseData object.
        row: Series
            Row of the DataFrame passed to __call__."""
        yield from ()  # pragma: no cover
